<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

**Мадина, привет!** 👋

Меня зовут Алексей Манелюк, я буду проверять твой проект. В Яндекс Практикуме принято общаться на "ты", но если некомфортно - скажи и перейдем на "Вы".

Моя основная цель - не указать на ошибки, а поделиться опытом и помочь в освоении анализа данных.

В тетрадке ты увидишь мои комментарии, пожалуйста, не удаляй, не перемещай и не изменяй их. У комментариев могут быть приписки **2**, **3** и так далее, которые соответствуют каждой версии проверки. Напомню, что у тебя **6** попыток для успешной сдачи проекта.

Я буду использовать цветовую разметку:
<p>
<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b>

Так выделены критические замечания. Без их отработки я не смогу принять проект. </div>

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>  Комментарий ревьюера ⚠️</b>

Так выделены небольшие замечания или полезные советы. Постарайся учесть их.</div>
<p>
<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Так выделены правильные решения.</div>

Предлагаю работать в диалоге: если что-то дорабатываешь или отвечаешь на мой комментарий - пиши об этом. Мне будет удобней, если ты выделишь свои комментарии:
<p>
<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Здесь исправил.</div>

Все это поможет мне быстрее принять твой проект.

<div class="alert-info" style="padding: 20px; border-radius: 10px">
<b>Комментарий студента 💎</b>

Ты можешь написать свой комментарий, если скопируешь этот блок в ячейку Markdown.</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Первая часть этого проекта проверяется автоматизрованно, но дам по ней пару советов - они пригодятся тебе в следующих проектах.

📌 Начать проект стоит с описания (цели исследования, описание данных и краткий план работы), а каждый раздел закончить промежуточным выводом.

Как правильно оформить проект ты можешь прочитать в материалах курса:

`Полезные инструкции для учёбы - Оформление проекта - Рекомендации по выполнению проектов`

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Определим функцию 
def mean(x):
    return '{:.2%} '.format(x.mean())
# Вызовем функцию
data.groupby('children')['debt'].agg(['count', mean])

count    mean
children               
0         14091  7.54% 
1          4808  9.23% 
2          2052  9.45% 
3           330  8.18% 
4            41  9.76% 
5             9  0.00%

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Это правильная сводная таблица, все нужные данные для сравнения категорий есть! Здорово, что написала функцию для форматирования столбца в процентный формат. 👍</div>

**Вывод:**  
- Заемщики без детей имеют наименьшую долю должников (7.54%).
- Доля должников среди заемщиков с одним ребенком выше и составляет 9.23%.
- Заемщики с двумя детьми также имеют высокую долю должников - 9.45%.

Таким образом, можно сделать вывод о том, что с увеличением количества детей у заемщика увеличивается вероятность задолженности по кредиту.

Однако группа с четыремя и пятью детьми требует дополнительного исследования, учитывая небольшой размер выборки.

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

Ты правильно обратила внимание на то, что выборки несбалансированы. Поэтому выявлять зависимость стоит, сравнивая доли должников в самых больших, сопоставимых по размеру категориях заемщиков. Как можно сформулировать зависимость, сравнив категории с 0, 1 и 2 детьми?</div>

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера 2 ✔️</b>

Если сравнить только самые многочисленные выборки с 0, 1 и 2 детьми, можно сделать вывод, что чем больше у заемщика детей, тем выше риск допустить просрочку.</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby(['family_status','family_status_id'])['debt'].agg(['count', mean]).sort_values('family_status_id')

count    mean
family_status         family_status_id               
женат / замужем       0                 12261  7.56% 
гражданский брак      1                  4134  9.31% 
вдовец / вдова        2                   951  6.62% 
в разводе             3                  1189  7.06% 
Не женат / не замужем 4                  2796  9.76%

**Вывод:** 
Да, действительно, наблюдается зависимость между семейным положением и возвратом кредита в срок.

Заемщики со статусом "Не женат / не замужем" имеют самую высокую долю задолженности по возврату кредитов - 9,76%, что является наивысшим показателем.

С другой стороны, заёмщики со статусом "вдовец / вдова" демонстрируют наименьшую долю задолженности - 6,62%. Это может быть связано с тем, что вдовцы и вдовы, как правило, более взрослые люди, обладающие большей ответственностью.

В общем, можно сказать, что наиболее безответственные в плане возврата кредита оказываются те, кто никогда не состоял в браке.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Верно, можно сказать, что состоящие или состоявшие в браке заемщики реже допускают просрочки.</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby(['total_income_category'])['debt'].agg(['count', mean]).sort_values('mean')

count    mean
total_income_category               
D                        349  6.02% 
B                       5014  7.06% 
A                         25  8.00% 
C                      15921  8.50% 
E                         22  9.09%

**Вывод:** 
Невозможно сделать однозначные выводы о зависимости между уровнем дохода и возвратом кредита в срок, так как размеры групп могут быть недостаточными для статистически значимого анализа. Тем не менее, сравнивая две самые многочисленные выборки, категории "B" и "C", можно заметить, что клиенты с более высоким уровнем дохода имеют более низкий процент задолженности по сравнению с клиентами с более низким уровнем дохода. Чем выше доход, тем ниже вероятность задолженности.

<div class="alert-success" style="border-radius: 10px; padding: 20px">
<b>Комментарий ревьюера 2 ✔️</b>

Все правильно!

<div class="alert-danger" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ❌</b></h5>

<s>А какую можно выявить зависимость, если сравнить только две самые многочисленные выборки?</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby(['purpose_category'])['debt'].agg(['count', mean]).sort_values('mean')

count    mean
purpose_category                       
операции с недвижимостью  10751  7.26% 
проведение свадьбы         2313  7.91% 
получение образования      3988  9.25% 
операции с автомобилем     4279  9.35%

**Вывод:** 
Клиенты, оформившие кредит на операции с недвижимостью, демонстрируют наилучший показатель возврата в срок - 7,26% невозврата кредита.
Клиенты, взявшие кредит на операции с автомобилем, являются наиболее безответственными, с 9,35% невозврата кредита.
Те, кто брал кредит на проведение свадьбы, показывают промежуточный результат - 7,91% невозврата.
Клиенты, взявшие кредит на получение образования, также имеют высокий процент невозврата - 9,25%.
Эти данные могут помочь банку принимать более обоснованные решения при выдаче кредитов, учитывая риски, связанные с разными целями займа.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Всё правильно. 👌</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Да, вероятно, пропуски в данных появились случайно. В большинстве случаев банки требуют заполнения всех обязательных полей заявки на кредит, таких как информация о доходах, семейном положении и других финансовых обязательствах, прежде чем одобрить заявку на кредит, особенно если речь идет о крупных суммах или кредите на недвижимость.

Тем не менее, возможны различные причины появления пропусков в данных, например, технические ошибки при вводе данных или неполнота предоставленной информации. Но в целом, предполагается, что заявки на кредит, в которых отсутствует важная информация, могли быть отклонены.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Предположения о причинах возникновении пропусков правильные: либо человеческий фактор, либо техническая ошибка. </div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана менее чувствительна к выбросам и отражает центральную тенденцию данных более устойчиво, чем среднее значение.

Использование медианы для заполнения пропущенных значений в столбцах данных может быть разумным решением, особенно если данные имеют смещенное распределение или содержат выбросы.

Однако следует помнить, что заполнение пропусков медианой по всему столбцу также может исказить данные, особенно если пропущенные значения не случайны и имеют систематический характер. В таких случаях более предпочтительным может быть использование более сложных методов обработки пропусков, таких как импутация с использованием моделей машинного обучения или анализа корреляции с другими признаками для более точного заполнения пропущенных значений.

<div class="alert-success" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ✔️</b>

Согласен.

### Шаг 4: общий вывод.

Общий вывод:
Проделанная работа по исследованию надежности заемщиков включала анализ полученных данных, предварительную обработку, заполнение пропусков, удаление дубликатов и объединение клиентов в группы по различным параметрам. На основе этой работы можно сделать следующие выводы:

1.Зависимость между количеством детей и возвратом кредита в срок: Бездетные заемщики демонстрируют лучшие показатели возврата кредита, что может быть связано с меньшими непредвиденными расходами. Однако для группы с пятью детьми нельзя сделать однозначных выводов из-за маленького размера выборки.

2.Зависимость между семейным положением и возвратом кредита в срок: Обнаружена зависимость между семейным положением и возвратом кредита. Заемщики, не состоящие в браке, имеют наибольший процент невозврата, в то время как вдовцы и вдовы возвращают кредиты лучше всего.

3.Зависимость между уровнем дохода и возвратом кредита в срок: Клиенты с более высоким уровнем дохода имеют более низкий процент задолженности по сравнению с клиентами с более низким уровнем дохода. Можно предположить, что чем выше доход, тем ниже вероятность задолженности.

4.Влияние целей кредита на его возврат в срок: Заемщики, берущие кредит на операции с недвижимостью, имеют наилучшие показатели возврата, в то время как те, кто берет кредит на операции с автомобилем, являются наиболее неплатежеспособными. Заемщики, берущие кредит на проведение свадьбы и получение образования, находятся между ними по уровню невозврата кредита.

В целом, анализ этих параметров поможет банку принимать более обоснованные решения при выдаче кредитов, учитывая финансовые риски, связанные с различными категориями заемщиков и целями кредита.

<div class="alert-warning" style="padding: 20px; border-radius: 10px">
<b>Комментарий ревьюера ⚠️</b>

<s>Я бы предложил немного поправить итоговый вывод:
    
- начать с абзаца с описания исходных данных и что было сделано на стадии предобработки
- скоррректировать зависимости просрочки от количества детей и уровня дохода (после доработки проекта)


<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

<h5><b>🏁 Общий комментарий ревьюера:</b></h5>

Ты хорошо поработала над проектом, в целом все круто, но нужно кое-что исправить, чтобы я его принял.

У меня только одно критическое замечание - касательно выявления зависимости просрочки от уровня дохода.

Также, постарайся отработать желтые комментарии.

Буду ждать твой доработанный проект. ⌛️</div>

<div style="border:solid steelblue 3px; padding: 20px; border-radius: 10px">

<h5><b>🏁 Общий комментарий ревьюера 2</b></h5>

Спасибо, что доработала проект! Критических замечаний больше нет, получилась очень хорошая работа.

Успехов в дальнейшем обучении! 🏆

</div>